pandas/tests/series/methods/test_fillna.py

```diff
@@ -623,6 +623,35 @@ def test_fillna_numeric_inplace(self):
         expected = x.fillna(value=0)
         tm.assert_series_equal(y, expected)
 
+    def test_fillna_null_value_replacement(self, frame_or_series):
```

Comment:
can you not use nulls_fixture here?

Reply:
If we wanted to be more thorough by testing all the permutations of nulls we would probably need to discuss a few things.  If we wrote the unit tests with all the unique null fixtures we run into several different problems:

```
def test_fillna_null_value_replacement(self, frame_or_series, unique_nulls_fixture, unique_nulls_fixture2):
      # GH#40498
      ser = Series([unique_nulls_fixture, unique_nulls_fixture])
      obj = frame_or_series(ser)

      value = unique_nulls_fixture2
      result = obj.fillna(value)

      expected = Series([unique_nulls_fixture2, unique_nulls_fixture2])
      expected = frame_or_series(expected)

      tm.assert_equal(result, expected)
```

> FAILED pandas/tests/series/methods/test_fillna.py::TestSeriesFillNA::test_fillna_null_value_replacement[DataFrame-None-None] - ValueError: Must specify a fill 'value' or 'method'.
> FAILED pandas/tests/series/methods/test_fillna.py::TestSeriesFillNA::test_fillna_null_value_replacement[DataFrame-None-unique_nulls_fixture22] - AssertionError: Attributes of DataFrame.iloc[:, 0] (column name="0") are different
> FAILED pandas/tests/series/methods/test_fillna.py::TestSeriesFillNA::test_fillna_null_value_replacement[DataFrame-nan-None] - ValueError: Must specify a fill 'value' or 'method'.
> FAILED pandas/tests/series/methods/test_fillna.py::TestSeriesFillNA::test_fillna_null_value_replacement[DataFrame-nan-unique_nulls_fixture22] - AssertionError: Attributes of DataFrame.iloc[:, 0] (column name="0") are different
> FAILED pandas/tests/series/methods/test_fillna.py::TestSeriesFillNA::test_fillna_null_value_replacement[DataFrame-unique_nulls_fixture2-None] - ValueError: Must specify a fill 'value' or 'method'.
> FAILED pandas/tests/series/methods/test_fillna.py::TestSeriesFillNA::test_fillna_null_value_replacement[DataFrame-unique_nulls_fixture2-nan] - AssertionError: Attributes of DataFrame.iloc[:, 0] (column name="0") are different
> FAILED pandas/tests/series/methods/test_fillna.py::TestSeriesFillNA::test_fillna_null_value_replacement[Series-None-None] - ValueError: Must specify a fill 'value' or 'method'.
> FAILED pandas/tests/series/methods/test_fillna.py::TestSeriesFillNA::test_fillna_null_value_replacement[Series-None-unique_nulls_fixture22] - AssertionError: Attributes of Series are different
> FAILED pandas/tests/series/methods/test_fillna.py::TestSeriesFillNA::test_fillna_null_value_replacement[Series-nan-None] - ValueError: Must specify a fill 'value' or 'method'.
> FAILED pandas/tests/series/methods/test_fillna.py::TestSeriesFillNA::test_fillna_null_value_replacement[Series-nan-unique_nulls_fixture22] - AssertionError: Attributes of Series are different
> FAILED pandas/tests/series/methods/test_fillna.py::TestSeriesFillNA::test_fillna_null_value_replacement[Series-unique_nulls_fixture2-None] - ValueError: Must specify a fill 'value' or 'method'.
> FAILED pandas/tests/series/methods/test_fillna.py::TestSeriesFillNA::test_fillna_null_value_replacement[Series-unique_nulls_fixture2-nan] - AssertionError: Attributes of Series are different

These are what I found / my thoughts:
1. Filling with None is Ambiguous - We would either need to drop the ValueError and accept None as a valid value or add a way to distinguish when a user explicitly wants to fillna with None vs. when they forgot to add a valid value.
2. Casting Data Type - When filling with NaT the Series or DataFrame will not automatically change data types, however when filling with NaN it will.  Do we want to say this is expected behavior and add to the test or should we fix the casting.
3. Can't Replace NaT with NaN - This one is interesting but you actually can't replace this even with the following.
```
ser = Series([np.datetime64('NaT')])
ser.loc[:] = np.nan
```

Reply:
There were some problems with the null fixtures and how the test was written:
1. Filling with None is Ambiguous - We would either need to drop the ValueError and accept None as a valid value or add a way to distinguish when a user explicitly wants to fillna with None vs. when they forgot to add a valid value.
2. Casting Data Type - When filling with NaT the Series or DataFrame will not automatically change data types, however when filling with NaN it will. Do we want to say this is expected behavior and add to the test or should we fix the casting.
3. Can't Replace NaT with NaN - We actually can't replace this even with the following simplified logic. I'm not sure if this is a bug or expected.
```
ser = Series([np.datetime64('NaT')])
ser.loc[:] = np.nan
```

I changed it to use an object series which I think makes more sense (since users can always use .astype to cast a series of nulls without fillna).  The first issue still remains so I've set it to skip for those scenarios.  What do you think?